beacon-chain/db/kv/blob.go

```diff
@@ -0,0 +1,182 @@
+package kv
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v4/config/params"
+	types "github.com/prysmaticlabs/prysm/v4/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/v4/encoding/bytesutil"
+	ethpb "github.com/prysmaticlabs/prysm/v4/proto/prysm/v1alpha1"
+	bolt "go.etcd.io/bbolt"
+	"go.opencensus.io/trace"
+)
+
+// SaveBlobSidecar saves the blobs for a given epoch in the sidecar bucket. When we receive a blob:
+//
+//  1. Convert slot using a modulo operator to [0, maxSlots] where maxSlots = MAX_BLOB_EPOCHS*SLOTS_PER_EPOCH
+//
+//  2. Compute key for blob as bytes(slot_to_rotating_buffer(blob.slot)) ++ bytes(blob.slot) ++ blob.block_root
+//
+//  3. Begin the save algorithm:  If the incoming blob has a slot bigger than the saved slot at the spot
+//     in the rotating keys buffer, we overwrite all elements for that slot.
+func (s *Store) SaveBlobSidecar(ctx context.Context, scs []*ethpb.BlobSidecar) error {
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.SaveBlobSidecar")
+	defer span.End()
+
+	if len(scs) == 0 {
+		return errors.New("nil or empty blob sidecars")
+	}
+	slot := scs[0].Slot
```

Comment:
Good questions. I asked them before on Slack and the response from @kasey was that we don't want to validate it in the db package. I would say otherwise, because we want to make sure the db doesn't get corrupted in any way.

Reply:
added validations 831517166e9e0e025467a7030dba7b164bd734a5

Reply:
I was more opposed to the idea of verifying that the number of sidecars was less than the current value of `MAX_BLOBS_PER_BLOCK` than checking the consistency between slot and root. But yeah generally I do think that the db code checking the semantics *within* a value being stored is going a little too far and mixing concerns, especially since we're behind an abstract data storage interface here (kv). As opposed to for instance checking the integrity between values that refer to each other across atomic writes, which would definitely be the db's job, to ensure referential integrity.

Reply:
context: @rkapka is referring to my response to this comment: `I think we should add some more validation though, like checking that the number of saved sidecars is not more than MaxBlobsPerBlock and that all of them have the same slot etc`.